### ISSUE TITLE
Switching to a pub-sub model for stock data

### ISSUE THREAD

**Issue by kirankaranth1** [vinyasmusic/price-alert#5]
Currently every single row in the DB is iterated upon and an API call made for each row. If many users want stock updates for the same symbols with similar/dissimilar filters this will result in too many unnecessary API calls.

Switching to pub-sub model  will solve this. There would be a separate publisher that just publishes data of each unique symbol that users are currently interested in. When a user wants to track a new stock a subscription would be created to track publishes to that stock with the user specified filter (specific percentage/value) etc.

 This will make the system very scalable. I'm assuming a normal message broker like redis can accomplish this.

**Comment by pyvinyas** [comment]
To achieve this I don't think a redis broker would be necessary as such. Here's what I think can be done which will require tinkering the data model. 

Currently I am creating a entry for each alert created by a user. I can change this to being a single entry for a scrip symbol eg. HDFC where we will store the subscribers details and filter in another model. This will be a OneToMany relation hence we only need to do the API call once for a symbol. After doing the API call we can get the list of users who have subscribed then check whether their condition is met or not. 
Also to ease this I can make the scrip symbol be a typeahead which gives user the choices as and when they type. 

Let me know your thoughts.

**Comment by kirankaranth1** [comment]
Yes, this is better on the same lines. I suggested pub-sub because the condition filtering will again have to happen in background every six hours for every single alert and that too can be reduced. A message broker will take care that a callback to the main service occurs only when the conditions are met (as that is how the subscription was created initially) so you can just go ahead with sending notification without any DB read/write.

**Comment by pyvinyas** [comment]
So if I understood it correctly : 
I can use the new model and then create a subscription for each new alert added. Whenever the condition is met the broker will send a signal to the email task for that user.

**Comment by kirankaranth1** [comment]
Yes.
